p2p/parts_assembler.go

```diff
@@ -0,0 +1,87 @@
+// Copyright 2016 Factom Foundation
+// Use of this source code is governed by the MIT
+// license that can be found in the LICENSE file.
+
+package p2p
+
+import (
+	"time"
+)
+
+// maximum time we wait for a partial message to arrive, old entries are cleaned up only when new part arrives
+const MaxTimeWaitingForReassembly time.Duration = time.Second * 60 * 10
+
+type PartialMessage struct {
+	parts                  []*Parcel // array of message parts we've received so far
+	firstPartReceived      time.Time // a timestamp indicating when the first part was received
+	mostRecentPartReceived time.Time // a timestamp indicating when the mostRecent part was received
+}
+
+// PartsAssembler is responsible for assembling message parts into full messages
+type PartsAssembler struct {
+	messages map[string]*PartialMessage // a map of app hashes to partial messages
+}
+
+// Initializes the assembler
+func (assembler *PartsAssembler) Init() *PartsAssembler {
+	assembler.messages = make(map[string]*PartialMessage)
+	return assembler
+}
+
+// Handles a single message part, returns either a fully assembled message or nil
+func (assembler *PartsAssembler) handlePart(parcel Parcel) *Parcel {
+	debug("PartsAssembler", "Handling message part %d %d/%d", parcel.Header.AppHash, parcel.Header.PartNo, parcel.Header.PartsTotal)
+	partial, exists := assembler.messages[parcel.Header.AppHash]
+
+	if !exists {
+		partial = createNewPartialMessage(parcel)
+		assembler.messages[parcel.Header.AppHash] = partial
+	}
+
+	partial.parts[parcel.Header.PartNo] = &parcel
+	partial.mostRecentPartReceived = time.Now()
+
+	fullParcel := tryReassemblingMessage(partial)
```

Comment:
For legibility, perhaps a comment here that `tryReassemblingMessage` returns the message or `nil`
(to map to the overall comment on L31)
It will avoid having to do a lookup in the code to understand this...